src/codecs/tiff.rs

```diff
@@ -296,24 +381,91 @@ fn cmyk_to_rgb(cmyk: &[u8]) -> [u8; 3] {
     ]
 }
 
-// Utility to simplify and deduplicate error handling during 16-bit encoding.
-fn u8_slice_as_u16(buf: &[u8]) -> ImageResult<&[u16]> {
-    bytemuck::try_cast_slice(buf).map_err(|err| {
-        // If the buffer is not aligned or the correct length for a u16 slice, err.
-        //
-        // `bytemuck::PodCastError` of bytemuck-1.2.0 does not implement
-        // `Error` and `Display` trait.
-        // See <https://github.com/Lokathor/bytemuck/issues/22>.
-        ImageError::Parameter(ParameterError::from_kind(ParameterErrorKind::Generic(
-            format!("{:?}", err),
-        )))
-    })
+enum DtypeContainer<'a, T> {
+    Slice(&'a [T]),
+    Vec(Vec<T>),
+}
+
+impl<T> DtypeContainer<'_, T> {
+    fn as_slice(&self) -> &[T] {
+        match self {
+            DtypeContainer::Slice(slice) => slice,
+            DtypeContainer::Vec(vec) => vec,
+        }
+    }
+}
+
+fn u8_slice_as_f32(buf: &[u8]) -> ImageResult<DtypeContainer<f32>> {
+    let res = bytemuck::try_cast_slice(buf);
+    match res {
+        Ok(slc) => Ok(DtypeContainer::<f32>::Slice(slc)),
+        Err(err) => {
+            match err {
+                bytemuck::PodCastError::TargetAlignmentGreaterAndInputNotAligned => {
+                    // If the buffer is not aligned for a f32 slice, copy the buffer into a new Vec<f32>
+                    let mut vec = vec![0.0; buf.len() / 4];
+                    for (i, chunk) in buf.chunks_exact(4).enumerate() {
+                        let f32_val = f32::from_ne_bytes([chunk[0], chunk[1], chunk[2], chunk[3]]);
+                        vec[i] = f32_val;
+                    }
+                    Ok(DtypeContainer::Vec(vec))
+                }
+                _ => {
+                    // If the buffer is not the correct length for a f32 slice, err.
+                    Err(ImageError::Parameter(ParameterError::from_kind(
+                        ParameterErrorKind::Generic(format!("{:?}", err)),
+                    )))
+                }
+            }
+        }
+    }
+}
```

Comment:
I'm surprised this is not a generic method bounded by `Pod`. In particular some of the verbosity around `from_ne_bytes` is simulating an unaligned read, a [method bytemuck provides](https://docs.rs/bytemuck/latest/bytemuck/fn.pod_read_unaligned.html) for the relevant sample types.